signature/signer.go

```diff
@@ -0,0 +1,130 @@
+package signature
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"errors"
+	"fmt"
+)
+
+// Signer is used to sign bytes generated after signature envelope created.
+type Signer interface {
+	// Sign signs the payload and returns the raw signature and certificates.
+	Sign(payload []byte) ([]byte, []*x509.Certificate, error)
+
+	// KeySpec returns the key specification.
+	KeySpec() (KeySpec, error)
+}
+
+// LocalSigner is used by built-in signers to sign only.
+type LocalSigner interface {
+	Signer
+
+	// CertificateChain returns the certificate chain.
+	CertificateChain() ([]*x509.Certificate, error)
+
+	// PrivateKey returns the private key.
+	PrivateKey() crypto.PrivateKey
+}
+
+// signer is a LocalSigner implementation.
```

Comment:
It is worth calling out that the `signer` only holds the signing key. The underlying signing implementation is provided by the underlying crypto library for the specific signature format like `go-jwt` and `go-cose`.